corehq/apps/reports/standard/cases/case_data.py

```diff
@@ -555,6 +559,209 @@ def close_case_view(request, domain, case_id):
     return HttpResponseRedirect(reverse('case_data', args=[domain, case_id]))
 
 
+MAX_CASE_COUNT = 10
+MAX_SUBCASE_DEPTH = 3
+
+
+@location_safe
+def get_cases_and_forms_for_deletion(request, domain, case_id):
+    delete_cases = []  # list of cases to be soft deleted
+    delete_forms = []  # list of forms to be soft deleted
+
+    # For formatting the list of cases/submission forms
+    cases = {}  # structured like: {case: {form: {case touched by form: actions taken by form}}}
```

Comment:
perhaps a dataclass or similar data structure could be used to give this better structure.

Reply:
@minhaminha I think something like a `namedtuple` would be helpful, here is an example:
https://github.com/dimagi/commcare-hq/blob/72e347c33397674503fcea10d4a67f0c1dea0a01/corehq/apps/cleanup/management/commands/undelete_docs.py#L13

you can then call `. _asdict()` on your `namedtuple`

Reply:
@millerdev I always found `namedtuple` nice due to the shorter notation for a quick dataclass-esq object for simple use cases like this. curious where your preference for a full attrs object comes in vs a `namedtuple`? Usually I would use the former if I wanted optional fields and methods, but `namedtuple` for very quick named attributes

Reply:
I find dataclasses and attrs to support more natural class syntax than `namedtuple`. There is tons more flexibility such as field defaults like `factory=list`, it's trivial to add methods, etc. Also, something is sticking in my mind that `namedtuple` has a performance overhead penalty, but I could be wrong about that. [SO also points out](https://stackoverflow.com/a/73724964/10840) that

>  When you compare named tuples, the names are ignored: two named tuples are equal if they contain the same values in the same order, even if they have different class names or field names

Reply:
One other thing is namedtuple doesn't support default values which is often useful.

Also, I find most IDEs quite bad at backlinking to namedtuple definitions since they are just a variable and have no static class definition.